src/Subsystem/ElevatorSubsytem/Elevator.java

```diff
@@ -108,33 +115,131 @@ private void sendStateUpdate(){
-     *
-     * @param event The event to process
-     * @throws InvalidTypeException If it receives an event type this class cannot handle
-     */
-    private void processMessage(SystemMessage event) throws InvalidTypeException {
-        // Note: Cannot switch on type, if we want to refactor selection, look into visitor pattern.
-        if (event instanceof MoveElevatorCommand)
-            move(((MoveElevatorCommand) event).direction());
-        else if (event instanceof MovePassengersCommand) {
-            unload();
-            load((MovePassengersCommand) event);
-        }
-        else // Default, should never happen
-            throw new InvalidTypeException("Event type received cannot be handled by this subsystem.");
-    }
-
     @Override
     public void run() {
+        // Set initial state.
+        //
+        // Decision to set initial state in method run because the elevator
+        // has no state until method run is invoked.
+        //
+        // Alternative would be setting initial state in constructor but this
+        // causes funny business with transmitters and receivers.
+        setState(new ReceivingState(this));
+
         while (true){
-            sendStateUpdate();
-            try {
-                processMessage(receiver.dequeueMessage());
-            } catch (InvalidTypeException e) {
+            // Wait until an event occurs in this state machine
+            event = receiver.dequeueMessage();
+
+            // Handle the event occurring in this state machine
+            if (event instanceof MoveElevatorCommand) {
+                ((ElevatorState) state).handleMoveElevatorCommand();
+            } else if (event instanceof MovePassengersCommand) {
+                ((ElevatorState) state).handleMovePassengersCommand();
+            } else {
+                InvalidTypeException e = new InvalidTypeException("Event type received cannot be handled by this subsystem.");
                 throw new RuntimeException(e);
             }
         }
     }
+
+    /**
+     * Abstract class representing an elevator state.
+     *
+     * @author Braeden Kloke
+     * @version March 5, 2024
+     */
+    public abstract class ElevatorState extends State {
+
+        public ElevatorState(Context context) {
+            super(context);
+        }
+
+        /**
+         * Handles state machine behaviour when event MoveElevatorCommand occurs.
+         *
+         * @author Braeden Kloke
+         */
+        public void handleMoveElevatorCommand() {}
+
+        /**
+         * Handles state machine behaviour when event MovePassengersCommand occurs.
+         *
+         * @author Braeden Kloke
+         */
+        public void handleMovePassengersCommand() {}
+    }
+
+    /**
+     * Class representing the state for an elevator receiving system messages.
+     *
+     * @author Braeden Kloke
+     * @version March 5, 2024
+     */
+    public class ReceivingState extends ElevatorState {
```

Comment:
factor out of the file

Reply:
this is where we should wait to receive the message, not in the elevator run(). basically in the entry method, we do the receiving method, then in the do activity we process it and select the next state

Reply:
Comment 1: Refactored. Done.

Comment 2: I don't agree with the suggested implementation. See bulleted list below for reasoning.

- Events act upon a state machine. In other words, an event happens and whatever state we are in has to handle it. 
- Events happen by being dequeued from the receiver.
- After an event happens, the current state in our state machine needs to immediately respond to the event.
- Thus, encapsulating 'events happening' all inside the ReceivingState would not be implementing a state machine because this implies events ONLY happen in the receiving state. This is false since events should be able to happen in any state.
- My implementation is more representative of an actual state machine where events have the possibility of happening in any state.

Reply:
I think I'm understanding the debate. The way I see it, Braeden has the Elevator thread doing the initial processing of the message popped off of the Receiver (ie. switching it as either the MoveElevatorCommand or MovePassengersCommand), and then calling the ReceivingState's method which then transitions to the appropriate state. I think Arthur is suggesting just having the run() method basically just set an initial state, and then only having the ReceivingState capable of popping/processing commands.

I think if we put the process message logic in the ReceivingState, that seems to make sense to me. I agree with Braeden's points about the state machine and that events can happen to it regardless of what state it is in; however, if a state has no transition for a certain event, nothing needs to happen. Again, I don't know if I'm understanding it right, but I think it is actually desired behaviour that only the ReceivingState is capable of getting the MoveElevatorCommand/MovePassengersCommand. Other events still happen in all the other states (eg. time-triggered event of load-time elapsing in load() -> like the time-to-live  TL=LOAD_TIME thing in Wainer's extended FSMs, if you remember that!; same thing with the move() command and TL=TRAVEL_TIME). 

For me, I was confused thinking that MoveElevatorCommand/MovePassengersCommand were the only events here; they're just the only events that are relevant when in the ReceivingState. If we're in a non-ReceivingState, I don't think we want to be popping anything off the queue - we just want our time-triggered event to fire, move us back to the ReceivingState, and then pop off queue.

So, I think the functionality is basically going to be the same either way, but yeah - maybe makes sense to only have MoveElevatorCommand/MovePassengersCommand pop/split in the ReceivingState. Might also be more in line with what we'll be doing for the other state machines.

P.S. Kindly disregard if this makes no sense.